search_log.go

```diff
@@ -136,34 +177,45 @@ func readLine(reader *bufio.Reader) (string, error) {
 	return string(line), nil
 }
 
-func readLastLine(file *os.File) string {
-	var line []byte
-	var cursor int64
-	stat, _ := file.Stat()
-	filesize := stat.Size()
+// Read lines from the end of a file
+// endCursor initial value should be the filesize
+func readLastLines(file *os.File, endCursor int64) string {
+	var lines []byte
+	var cursor = endCursor
 	for {
-		cursor -= 1
-		file.Seek(cursor, io.SeekEnd)
-
-		char := make([]byte, 1)
-		file.Read(char)
-
-		// stop if we find a line
-		if cursor != -1 && (char[0] == 10 || char[0] == 13) {
+		// stop if we are at the begining
+		// check it in the start to avoid read beyond the size
+		if cursor <= 0 {
 			break
 		}
-		line = append(line, char[0])
-		if cursor == -filesize { // stop if we are at the begining
-			break
+
+		var size int64 = 512
+		if cursor < size {
+			size = cursor
+		}
+		cursor -= size
+
+		file.Seek(cursor, io.SeekStart)
+		chars := make([]byte, size)
+		file.Read(chars)
+		lines = append(chars, lines...)
+
+		// find first '\n' or '\r'
+		for i := 0; i < len(chars); i++ {
```

Comment:
We only need to check `\r` and `\n` in the scanned bytes from this round, instead of previous rounds

Reply:
Yes, it works as you said, `len(chars)` is the length of content just read in this round.